File: Tree/Count_Good_Nodes_in_Binary_Tree.py
#Problem Link:- https://leetcode.com/problems/count-good-nodes-in-binary-tree/
"""Given a binary tree root, a node X in the tree is named good if in the path from root to 
X there are no nodes with a value greater than X.Return the number of good nodes in the 
binary tree."""

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if root is None:
            return 0
        self.ans = 0
        def helper(root,mx):
            if root is None:
                return 0
            if root.val >= mx:
                self.ans += 1
                mx = root.val
            helper(root.left,mx)
            helper(root.right,mx)
        helper(root,float("-inf"))
        return self.ans

File: Tree/CountGoodNodesInBinaryTree.cpp
//Problem Link:- https://leetcode.com/problems/count-good-nodes-in-binary-tree/

#include<bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
 
class Solution {
public:
    int helper(TreeNode* root, int mx){
        if (!root){
            return 0;
        }
        if (root->val >= mx){
            mx = root->val;
            return 1 + helper(root->left,mx) + helper(root->right,mx);
        }
        return helper(root->left,mx) + helper(root->right,mx);
    }
    
    int goodNodes(TreeNode* root) {
        if (!root){
            return 0;
        }
        if (!root->left && !root->right){
            return 1;
        }
        return 1 + helper(root->left,root->val) + helper(root->right,root->val);
    }
};